[outliners] Turn nooutline into an Enum Attribute

lenary opened the pull request:
This change turns the `"nooutline"` attribute into an enum attribute
called `nooutline`, and adds an auto-upgrader for bitcode to make the
same change to existing IR.

This IR attribute disables both the Machine Outliner (enabled at Oz for
some targets), and the IR Outliner (disabled by default).

lenary committed:
[outliners] Turn nooutline into an Enum Attribute

This change turns the `"nooutline"` attribute into an enum attribute
called `nooutline`, and adds an auto-upgrader for bitcode to make the
same change to existing IR.

This IR attribute disables both the Machine Outliner (enabled at Oz for
some targets), and the IR Outliner (disabled by default).

lenary committed:
Release Note

topperc commented on llvm/test/Bitcode/upgrade-nooutline.ll:
```diff
@@ -0,0 +1,12 @@
+; RUN: llvm-as < %s | llvm-dis - | FileCheck %s
+
+; CHECK: define void @f() [[ATTR:#[0-9]+]]
```
Why do we need so many NOT? Can we check the full final attribute list? Are other unrelated attributes being added?

topperc commented:
Should we proactively update the various syntax highlighting files in the utils directory?

llvm/utils/kate/llvm.xml
llvm/utils/vim/syntax/llvm.vim
llvm/utils/vscode/llvm/syntaxes/ll.tmLanguage.yaml

lenary commented on llvm/lib/IR/AutoUpgrade.cpp:
```diff
@@ -6345,6 +6345,12 @@ void llvm::UpgradeFunctionAttributes(Function &F) {
 
+  if (Attribute A = F.getFnAttribute("nooutline");
+      A.isValid() && A.isStringAttribute()) {
+    F.removeFnAttr("nooutline");
```
Switched to this. Update coming.

lenary committed:
Code Review Updates:

- Use AttrsToAdd/AttrsToRemove
- Use --implict-check-not in the test
- Change LLVM IR editor languages
- Propagage NoOutline in CodeExtractor (as `"nooutline"` was)

topperc reviewed (approved):
LGTM

lenary commented:
### Merge activity

* **Feb 11, 5:42 AM UTC**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.com/github/pr/llvm/llvm-project/163665).
* **Feb 11, 5:44 AM UTC**: @lenary merged this pull request with [Graphite](https://app.graphite.com/github/pr/llvm/llvm-project/163665).
